### ISSUE TITLE
Camera Light opening but video preview not available.

### ISSUE THREAD

**Issue by Sagar1219** [aws/amazon-chime-sdk-js#1460]
### What happened and what did you expect to happen?

I am using chime SDK for JS to integrate into the client application & in the video preview section, I am unable to see the video being played, although camera light starts to blink. There is no permission issue. Have gone through the complete README and Getting Started Documentation but this issue is still not resolving.

### Have you reviewed our existing documentation?

- [X] GitHub [issues](https://github.com/aws/amazon-chime-sdk-js/issues)
- [X] FAQs [page](https://aws.github.io/amazon-chime-sdk-js/modules/faqs.html)
- [X] README [page](https://github.com/aws/amazon-chime-sdk-js#amazon-chime-sdk-for-javascript)
- [X] Feature-specific [guides](https://github.com/aws/amazon-chime-sdk-js#guides)

### Reproduction steps

This is the HTML :
```
<div class="col-sm-6 text-center d-sm-block video-preview">
<video id="video-preview" class="w-100 h-100" style="max-width:360px;max-height:240px;border-radius:8px"></video>
</div>
```

This is the JS code that I have added : 
audioVideo.setDeviceLabelTrigger(() => Promise.resolve(new MediaStream()));

// Invoke devices
audioVideo.setDeviceLabelTrigger(async () =>
  await navigator.mediaDevices.getUserMedia({ audio: true, video: true })
);
audioVideo.chooseVideoInputDevice(videoInputAsync[0].deviceId);
audioVideo.start();

### Amazon Chime SDK for JavaScript version

2.13.0

### What browsers are you seeing the problem on?

Chrome

### Browser version

Version 91.0.4472.164 (Official Build) (64-bit)

### Browser console logs

2021-07-26T12:35:27.892Z [INFO] ChimeMeetingLogs - API/DefaultAudioVideoFacade/d9f9a87c-71fe-4bad-81f7-ed59167a0706/29c0b2d1-67ac-1ef8-3f5b-0eb354463602/listVideoInputDevices null -> [{"deviceId":"9f6c94f217f28d14834b6a2da61c3589d3ec28a4e5915b44da3a1e8a9dcb0a98","kind":"videoinput","label":"Integrated Camera (04ca:7070)","groupId":"5a829ebc8b5f70cb29874843d23ce8710b1005d8978b2036b26ad8dcbec9bfa0"}]

2021-07-26T12:35:27.892Z [INFO] ChimeMeetingLogs - API/DefaultAudioVideoFacade/d9f9a87c-71fe-4bad-81f7-ed59167a0706/29c0b2d1-67ac-1ef8-3f5b-0eb354463602/chooseVideoInputDevice "9f6c94f217f28d14834b6a2da61c3589d3ec28a4e5915b44da3a1e8a9dcb0a98"
ConsoleLogger.js:68 

2021-07-26T12:35:27.892Z [INFO] ChimeMeetingLogs - video send has ideal max bandwidth 1400 kbps
ConsoleLogger.js:68 

2021-07-26T12:35:27.892Z [INFO] ChimeMeetingLogs - requesting new video device with constraint {"video":{"deviceId":{"exact":"9f6c94f217f28d14834b6a2da61c3589d3ec28a4e5915b44da3a1e8a9dcb0a98"},"width":{"ideal":960},"height":{"ideal":540},"frameRate":{"ideal":15},"googCpuOveruseDetection":true,"googCpuOveruseEncodeUsage":true,"googCpuOveruseThreshold":85,"googCpuUnderuseThreshold":55}}

2021-07-26T12:35:29.387Z [INFO] ChimeMeetingLogs - AudioVideoStart/d9f9a87c-71fe-4bad-81f7-ed59167a0706/29c0b2d1-67ac-1ef8-3f5b-0eb354463602/Timeout15000ms/Media/Signaling/JoinAndReceiveIndexTask took 666 ms

2021-07-26T12:35:29.393Z [INFO] ChimeMeetingLogs - attaching audio track to peer connection (unified-plan)
ConsoleLogger.js:68 

2021-07-26T12:35:29.394Z [INFO] ChimeMeetingLogs - peer connection negotiation is needed
ConsoleLogger.js:68 

2021-07-26T12:35:29.395Z [INFO] ChimeMeetingLogs - no video track

2021-07-26T12:35:29.975Z [INFO] ChimeMeetingLogs - received new index {"atCapacity":false,"numParticipants":1}
ConsoleLogger.js:68

 2021-07-26T12:35:29.975Z [INFO] ChimeMeetingLogs - should resubscribe: false (downlink: false uplink: false)

2021-07-26T12:35:29.977Z [INFO] ChimeMeetingLogs - set remote description, waiting for ICE connection
ConsoleLogger.js:68

2021-07-26T12:35:29.990Z [INFO] ChimeMeetingLogs - peer connection state changed: connecting
ConsoleLogger.js:68 

2021-07-26T12:35:30.133Z [INFO] ChimeMeetingLogs - peer connection ice connection state changed: connected

**Comment by simmkyu** [comment]
The Chime SDK for JavaScript offers the `meetingSession.audioVideo.startVideoPreviewForVideoInput` API for previewing your video: https://aws.github.io/amazon-chime-sdk-js/interfaces/audiovideofacade.html#startvideopreviewforvideoinput

Sorry for the confusion that the README does not include a usage example for this API. You can find an example in the Chime SDK meeting demo. Before starting a meeting (`meetingSession.audioVideo.start` API), choose a video input device and then call the `startVideoPreviewForVideoInput` API to preview your video. Also, ensure that you call the [stopVideoPreviewForVideoInput API](https://aws.github.io/amazon-chime-sdk-js/interfaces/audiovideofacade.html#stopvideopreviewforvideoinput) to stop a video preview before joining a meeting.

https://github.com/aws/amazon-chime-sdk-js/blob/d7f8e3446a4bdec8adc8090039cb25fdad924853/demos/browser/app/meetingV2/meetingV2.ts#L2360-L2371

https://github.com/aws/amazon-chime-sdk-js/blob/d7f8e3446a4bdec8adc8090039cb25fdad924853/demos/browser/app/meetingV2/meetingV2.ts#L759-L762

**Comment by Sagar1219** [comment]
Hey, I was already using audioVideo.startVideoPreviewForVideoInput() yet it's not working.
This is the code for video previweing,
```

const videoInputDevices = audioVideo.listVideoInputDevices();
videoInputDevices.then((videoInputAsync) => {
      //List the devices in the dropdown
      var option = $('<option/>');
      option.attr({ 'value': videoInputAsync[0].deviceId }).text(videoInputAsync[0].label);  //Just first one only for now.
      $('#video-input').append(option);
      $('#video-input').val(videoInputAsync[0].deviceId);
      try {
        audioVideo.chooseVideoInputDevice(videoInputAsync[0].deviceId); 
      } catch (error) {
        console.log('failed to chooseVideoInputDevice', error); 
      }
    });
audioVideo.startVideoPreviewForVideoInput(document.getElementById('video-preview'));
```
It's not catching any error from the try catch statement, but got this in the console log : 

```
2021-07-27T06:05:43.874Z [WARN] ChimeMeetingLogs - cannot bind video preview since video input device has not been chosen
log @ ConsoleLogger.js:62
warn @ ConsoleLogger.js:34
startVideoPreviewForVideoInput @ DefaultDeviceController.js:221
startVideoPreviewForVideoInput @ DefaultAudioVideoFacade.js:269
initializeMeetingSession @ chime_js.js:111
(anonymous) @ chime_js.js:17
dispatch @ jquery.min.js:3
i @ jquery.min.js:3
```

Yet again, the camera light is opened but no preview available.

**Comment by devalevenkatesh** [comment]
Hello @Sagar1219 ,

From the pasted code, you call `audioVideo.startVideoPreviewForVideoInput(document.getElementById('video-preview'));` right after the `videoInputDevices.then` call. Hence, due to the asynchronous nature, the `audioVideo .startVideoPreviewForVideoInput` runs before `audioVideo.chooseVideoInputDevice` API and hence you are running into the "cannot bind video preview" error log.

For start video preview call to succeed could you please try calling that method after the `chooseVideoInputDevice` call and let us know if you are still running into any error.

```js
const videoInputDevices = await audioVideo.listVideoInputDevices();
//List the devices in the dropdown
var option = $('<option/>');
option.attr({ 'value': videoInputAsync[0].deviceId }).text(videoInputAsync[0].label);  //Just first one only for now.
$('#video-input').append(option);
$('#video-input').val(videoInputAsync[0].deviceId);
try {
  await audioVideo.chooseVideoInputDevice(videoInputAsync[0].deviceId); 
  // Added the call here
  audioVideo.startVideoPreviewForVideoInput(document.getElementById('video-preview'));
} catch (error) {
  console.log('failed to chooseVideoInputDevice', error); 
}
```

**Comment by Sagar1219** [comment]
Hi @devalevenkatesh 

Thanks for the response. I added the snippet that you shared, and I am getting the below error:

`
chime_js.js:72 Uncaught SyntaxError: await is only valid in async functions and the top level bodies of modules
`

This error is coming at this line: 
`
const videoInputDevices = await audioVideo.listVideoInputDevices();
`

Things to consider here: I am using the [min.js ](https://github.com/aws/amazon-chime-sdk-js/tree/master/demos/singlejs)functionality to load the SDK and then including it in the ` <script> ` tag. Is it because of this?

Also, in order to work a way around for this, I tweaked my code as :
```
videoInputDevices.then((videoInputAsync) => {
     	//List the devices in the dropdown
	var option = $('<option/>');
      	option.attr({ 'value': videoInputAsync[0].deviceId }).text(videoInputAsync[0].label);   //Just first one only for now.
      	$('#video-input').append(option);
      	$('#video-input').val(videoInputAsync[0].deviceId);
    })
    .then(()=>{
      	try {
        	audioVideo.chooseVideoInputDevice($('#video-input').val());
      	} catch (error) {
        	console.log(`failed to chooseVideoInputDevice`, error); 
      	}
    })
    .then(()=>{
      	audioVideo.startVideoPreviewForVideoInput(document.getElementById('video-preview'));
    });
```

Please help in finding a solution here.

**Comment by devalevenkatesh** [comment]
Sorry I assumed you would change the code snippet to an `async-await` function. So we have two options here:
1. Wrap the provided snippet in an `async` function. You can check the `async-await` pattern in the MDN docs [here](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await). For this, you have to change parent functions to be `async` since the `await` keyword can only be called in `async` functions. This is not the error from JS SDK but this is syntax error in JavaScript.
2. The `audioVideo.chooseVideoInputDevice` returns a promise hence, you can use `then` to know once it resolves and call `startVideoPreviewForVideoInput` in the `then` block. Sample code snippet:
```js
audioVideo.listVideoInputDevices()
.then((videoInputAsync) => {
    // Check here if you correctly get the video input devices.
    console.log('Printing the video input devices listed', videoInputAsync);
    // List the devices in the dropdown.
    var option = $('<option/>');
    option.attr({ 'value': videoInputAsync[0].deviceId }).text(videoInputAsync[0].label);   //Just first one only for now.
    $('#video-input').append(option);
    $('#video-input').val(videoInputAsync[0].deviceId);
    // This returns a Promise, hence, call start video preview after it completes.
    audioVideo.chooseVideoInputDevice($('#video-input').val())
    .then(() => {
        audioVideo.startVideoPreviewForVideoInput(document.getElementById('video-preview'));
    })
    .catch((err) => {
        console.log('Failed to chooseVideoInputDevice', err);
    });
});
```

Let us know if you are still running into issues.

**Comment by Sagar1219** [comment]
Hi @devalevenkatesh , 

My issue has been resolved. I have a similar issue with audio mute/unmute as well. Will apply this approach & if the issue still persists, then will open another issue ticket for that one. 

Thanks for the help. Closing this one now.

**Comment by ashirkhan94** [comment]
Hi @devalevenkatesh 
Same issue in React native side 
[https://github.com/aws-samples/amazon-chime-react-native-demo/issues/198](url)
Is there any solution for this in react native side 
Thanks.
